sdk/samples/iot/hub/src/paho_iot_hub_c2d_example.c

```diff
@@ -290,39 +321,63 @@ static int subscribe()
   return MQTTCLIENT_SUCCESS;
 }
 
-static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+static void receive_messages()
 {
-  (void)context;
   int rc;
-
-  printf("Received a message from service.\n");
-  printf("Topic: %s\n", topicName);
-
-  if (topicLen == 0)
-  {
-    // The length of the topic if there are one or more NULL characters embedded in topicName,
-    // otherwise topicLen is 0.
-    topicLen = (int)strlen(topicName);
-  }
-  az_span topic_span = az_span_init((uint8_t*)topicName, topicLen);
-
-  // Parse the incoming message topic and check to make sure it is a C2D message
+  char* topic = NULL;
+  int topic_len = 0;
+  MQTTClient_message* message = NULL;
   az_iot_hub_client_c2d_request c2d_request;
-  if (az_failed(rc = az_iot_hub_client_c2d_parse_received_topic(&client, topic_span, &c2d_request)))
-  {
-    printf("Topic is not a C2D message, az_result return code %04x\n", rc);
-  }
-  else
+
+  // Wait until max # messages received or timeout to receive a single message expires.
+  for(uint8_t message_count = 0; message_count < MAX_MESSAGE_COUNT; message_count++)
   {
-    if (message->payloadlen)
+    if (((rc
+          = MQTTClient_receive(mqtt_client, &topic, &topic_len, &message, TIMEOUT_MQTT_RECEIVE_MS))
+          != MQTTCLIENT_SUCCESS)
+          && (MQTTCLIENT_TOPICNAME_TRUNCATED != rc))
+    {
+      LOG_ERROR("Failed to receive message: MQTTClient return code %d.", rc);
+      exit(rc);
+    }
+    else if (NULL == message)
+    {
+      LOG_ERROR("Timeout expired: MQTTClient return code %d.", rc);
+      exit(rc);
+    }
+    else if (MQTTCLIENT_TOPICNAME_TRUNCATED == rc)
     {
-      printf("Received payload:\n%.*s\n", message->payloadlen, (char*)message->payload);
+      topic_len = (int)strlen(topic);
     }
+    LOG_SUCCESS("Message #%d: Client received message from the service.", message_count + 1);
+
+    parse_message(topic, topic_len, message, &c2d_request);
+    LOG_SUCCESS("Client parsed message.\n");
   }
 
-  putchar('\n');
   MQTTClient_freeMessage(&message);
-  MQTTClient_free(topicName);
+  MQTTClient_free(topic);
+  return;
+}
+
+static void parse_message(
+    char* topic,
+    int topic_len,
+    const MQTTClient_message* message,
+    az_iot_hub_client_c2d_request* c2d_request)
+{
+  int rc;
+  az_span topic_span = az_span_init((uint8_t*)topic, topic_len);
+  az_span message_span = az_span_init((uint8_t*)message->payload, message->payloadlen);
+
+  if (az_failed(rc = az_iot_hub_client_c2d_parse_received_topic(&client, topic_span, c2d_request)))
+  {
+    LOG_ERROR("Message from unknown topic: az_result return code 0x%04x.", rc);
+    exit(rc);
+  }
+  LOG_SUCCESS("Client received a valid topic response:");
+  LOG_AZ_SPAN("Topic:", topic_span);
+  LOG_AZ_SPAN("Payload:", message_span);
 
-  return 1;
+  return;
```

Comment:
nit: remove the unnecessary `return;` statement.
```suggestion
```

Reply:
Just to clarify, if you are fixing this up, unnecessary return statements for void-returning methods show up in a few places in the PR. We should fix them all :)